---
title: "QC Report"
author: 
date: 
output: html_document
params:
  samples: bs
  outdir: outdirectory
  organism: organism
  genome: genome
  reads: readData
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Rendered by scmeth

```{r,echo=FALSE,warning=FALSE,message=FALSE}
bs<-params$samples
nSamples<-length(Biobase::sampleNames(bs))
```

## Read information
#### If read data provided the following plot is shown
```{r,echo=FALSE,warning=FALSE,message=FALSE,fig.height=nSamples*0.4}
# Methylation distribution for all the cells
readinfo<-params$reads
if (!(is.null(readinfo))){
  scmeth::readmetrics(readinfo)
}


```


## CpG Coverage 

```{r,echo=FALSE,warning=FALSE,message=FALSE,fig.height=nSamples*0.4}

covVec<-scmeth::coverage(bs)
sampleNames<-bsseq::sampleNames(bs)
covDf <- data.frame(sample=sampleNames, coveredCpgs=covVec)
covDf$grp <- substr(covDf$sample, 1, 1)
m <- reshape2::melt(covDf, id.vars = c("grp", "sample"))
o <- order(covDf$grp, covDf$coveredCpgs)
sampleOrder <- covDf$sample[o]
m$sample <- factor(m$sample, levels=sampleOrder)

summary<-c('min','max','mean','median')
numericValues<-c(min(m$value),max(m$value),mean(m$value),median(m$value))
summaryDF<-as.data.frame(cbind(summary,numericValues))



ggplot2::ggplot(m, ggplot2::aes(sample, value)) + ggplot2::geom_hline(data = m, ggplot2::aes(yintercept = as.numeric(median(value))), linetype=2, color="blue")+ ggplot2::geom_bar(stat="identity") + ggplot2::coord_flip() + ggplot2::scale_y_continuous(name="Number of CpGs") + ggplot2::xlab("Sample") + ggplot2::ggtitle("Number of Covered CpGs")  + ggplot2::theme_bw()

#############################
```

## Bisulfite conversion rate for the samples
```{r,echo=FALSE,warning=FALSE,message=FALSE,fig.height=nSamples}
scmeth::bsConversionPlot(bs)

```


## Methylation Density plot
```{r,echo=FALSE,warning=FALSE,message=FALSE,fig.height=nSamples}
cpgDenMatrix<-scmeth::cpgDensity(bs,organism = params$organism,windowLength=1000)
cpgDenMatrix<-apply(cpgDenMatrix,1,function(x) x/colSums(cpgDenMatrix))
m <- reshape2::melt(cpgDenMatrix, varnames = c("Sample", "Density"))


ggplot2::ggplot(m, ggplot2::aes_string('Density','value')) + ggplot2::geom_line(ggplot2::aes_string(group='Sample')) + ggplot2::ggtitle("Fraction of covered CpGs by CpG density") + ggplot2::xlab("CpG coverage based on the CpG density") + ggplot2::xlab('CpG Density')+ggplot2::ylab('Fraction of CpGs')
#+ggplot2::theme(axis.text.x = element_text(angle = 60, hjust = 1))+
#ggplot2::theme(plot.title = element_text(hjust = 0.5))

```



## Non-repatmasker Coverage
```{r,echo=FALSE,warning=FALSE,message=FALSE,comment=FALSE,fig.height=nSamples*0.4}
covDf <- scmeth::repMask(bs,params$organism, params$genome)
covDf$sample<-rownames(covDf)
m <- reshape2::melt(covDf, id.vars = c("sample","coveredCpgs"))
o <- order(covDf$coveredCpgs)
sampleOrder <- covDf$sample[o]
m$sample <- factor(m$sample, levels=sampleOrder)

ggplot2::ggplot(m, ggplot2::aes_string('sample', 'coveredCpgs')) +  ggplot2::geom_bar(stat="identity") + ggplot2::coord_flip() + ggplot2::scale_y_continuous(name="Number of CpGs") + ggplot2::xlab("Sample") + ggplot2::ggtitle("Number of Covered Non-repeatmasker CpGs")  + ggplot2::theme_bw()


## Chromosomal Coverage
```

# Data Preprocessing
## CpG Discretization

```{r,echo=FALSE,warning=FALSE,message=FALSE,fig.height=nSamples*0.4}
# CpG Discretization
discretizedCpG<-cpgDiscretization(bs)
percentDiscarded<-discretizedCpG[3]
discardedDf <- data.frame(sample=sampleNames, value=percentDiscarded)
meltedDiscardedDf<- reshape2::melt(discardedDf, id.vars = c("sample", "discard.perc"))

o <- order(meltedDiscardedDf$discard.perc,meltedDiscardedDf$sample)
sampleOrder <- meltedDiscardedDf$sample[o]
meltedDiscardedDf$sample <- factor(meltedDiscardedDf$sample, levels=sampleOrder)

ggplot2::ggplot(meltedDiscardedDf, ggplot2::aes(sample, discard.perc)) +ggplot2::geom_bar(stat="identity") + ggplot2::coord_flip() + ggplot2::scale_y_continuous(name="Percentage of discarded CpGs") + ggplot2::xlab("Sample") + ggplot2::ggtitle("Discarded CpGs from each sample after discretization")  + ggplot2::theme_bw()

# Feature level coverage
#pandoc.header("Feature level coverage of the CpGs", level = 2)
library(annotatr)
featureCovDf<-scmeth::featureCoverage(bs,c('genes_promoters','genes_exons','genes_introns','genes_intergenic','cpg_islands','cpg_shelves','cpg_shores','cpg_inter'),genome)
colnames(featureCovDf)<-c('feature annotation','number of CpGs')
print(featureCovDf)
```

## Downsampling Analysis
```{r,echo=FALSE,warning=FALSE,message=FALSE,fig.height=nSamples}

##############################
# Downsample Plots
dsMatrix<-scmeth::downsample(bs)
meltedDsMatrix<-reshape2::melt(dsMatrix)
ggplot2::ggplot(meltedDsMatrix, ggplot2::aes(Var1, value, group=Var2))+ ggplot2::geom_line() + ggplot2::scale_x_continuous(name="Fraction of Reads") + ggplot2::scale_y_continuous(name="Number of CpG covered") + ggplot2::ggtitle("Saturation Analysis Plot") + ggplot2::theme_bw()

```

## Methylation Distribution
```{r,echo=FALSE,warning=FALSE,message=FALSE,fig.height=nSamples}
# Methylation distribution for all the cells
scmeth::methylationDist(bs,all=TRUE)

```


